### PR TITLE
[fixes #18139381] import csv & xls; export xls; move encoding to module

### PR DESCRIPTION
**Pull request by sachinr**
add specs; make date_parser more robust

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -20,6 +20,7 @@ gem 'rails', '2.3.12'
 gem 'rdoc'
 gem 'sass', '=3.1.4'
 gem 'settingslogic'
+gem "spreadsheet", "~> 0.6.5.9"
```
Comment (glennr) [inline Gemfile]: Are you locking the gem at the 0.6.5.x  intentionally ?  (  ~> locks it to least significant digit... )

File: app/controllers/application_controller.rb [hunk 2]
```diff
@@ -38,6 +38,12 @@ class ApplicationController < ActionController::Base
                 :disposition => "attachment; filename=#{filename}"
     end
 
+    def send_xls(data, filename)
```
Comment (glennr) [inline app/controllers/application_controller.rb:39]: please factor out common pattern send_csv() and send_xls()

File: app/helpers/application_helper.rb [hunk 3]
```diff
@@ -321,4 +321,13 @@ module ApplicationHelper
     end
     edit_activity_or_ocost_path(activity, :mode => mode)
   end
+
+  def sanitize_encoding(string)
+    begin
+      Iconv.conv("UTF-8//TRANSLIT//IGNORE", "UTF8", string)
+    rescue
+      Iconv.conv("UTF-8//TRANSLIT//IGNORE", "UTF8", string + ' ')[0..-1] # //IGNORE only ignores invalid byte sequences
```
Comment (glennr) [inline app/helpers/application_helper.rb]: line too long with comment - move comment above statement

File: app/models/importer.rb [hunk 4]
```diff
@@ -7,7 +7,13 @@ class Importer
   def initialize(response, filename)
     @response = response
     @filename = filename
-    @file = FasterCSV.open(@filename, {:headers => true, :skip_blanks => true})
+    begin
```
Comment (glennr) [inline app/models/importer.rb]: factor out the file opening into another method

File: app/models/importer.rb [hunk 5]
```diff
@@ -7,7 +7,13 @@ class Importer
...
+      worksheet = Spreadsheet.open(@filename).worksheet(0)
+      @file = create_hash_from_header(worksheet)
+    rescue Ole::Storage::FormatError
+      # try import the file as a csv if it is not an spreadsheet
```
Comment (glennr) [inline app/models/importer.rb]: what happens if its not a csv? how do we handle the exception. (at the very least we should document the expected behaviour in a spec - i.e. fastercsv should throw an (unhandled) exception)

File: app/helpers/application_helper.rb [hunk 6]
```diff
@@ -321,4 +321,13 @@ module ApplicationHelper
     end
     edit_activity_or_ocost_path(activity, :mode => mode)
   end
+
+  def sanitize_encoding(string)
```
Comment (glennr) [inline app/helpers/application_helper.rb]: this is a duplicate of lib/encoding_helper

remove this redundant method

File: lib/encoding_helper.rb [hunk 7]
```diff
@@ -0,0 +1,10 @@
+module EncodingHelper
+  def self.sanitize_encoding(string)
+    begin
+      Iconv.conv("UTF-8//TRANSLIT//IGNORE", "UTF8", string)
+    rescue
+      Iconv.conv("UTF-8//TRANSLIT//IGNORE", "UTF8", string + ' ')[0..-1] # //IGNORE only ignores invalid byte sequences
```
Comment (glennr) [inline lib/encoding_helper.rb]: line too long

File: spec/models/importer_spec.rb [hunk 8]
```diff
@@ -59,6 +59,56 @@ EOS
     end
   end
 
+  describe 'API (xls)' do
```
Comment (glennr) [inline spec/models/importer_spec.rb]: describe 'importing excel files'

File: app/models/importer.rb (hunk 5 above)
Comment (sachinr) [inline app/models/importer.rb]: If there is something wrong with the csv or it isn't a csv, a FasterCSV::MalformedCSVError is raised.  This is handled in the projects controller:

```
rescue FasterCSV::MalformedCSVError
  flash[:error] = "There was a problem with your file. Did you use the template provided and save the file as either XLS or CSV?
                   Please post a problem at <a href='https://hrtapp.tenderapp.com/kb'>TenderApp</a> if you can't figure out what's wrong."
  redirect_to response_projects_url(@response)
end
```

And tested in can_import_files.feature

**Pull request merged**